decoder.go

```diff
@@ -454,39 +456,79 @@ func (d *decoder) pointer(name string, value reflect.Value, def string) error {
 	return d.unmarshal(name, value.Elem(), def)
 }
 
-// Sets value to a channel type.
-func (d *decoder) channel(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
-
-// Sets value to a function type.
-func (d *decoder) function(key string, value reflect.Value, def string) error {
-	return d.textUnmarshaller(key, value, def)
-}
+// JSON encoder will fail to serialize maps that don't have strings as keys
+// so we are going to stringify them manually.
+func jsonMap(v interface{}) interface{} {
+	if reflect.TypeOf(v).Kind() == reflect.Map {
+		tmp := make(map[string]interface{})
+		rv := reflect.ValueOf(v)
+		for _, key := range rv.MapKeys() {
+			tmp[fmt.Sprint(key.Interface())] = jsonMap(rv.MapIndex(key).Interface())
+		}
 
-func (d *decoder) textUnmarshaller(key string, value reflect.Value, str string) error {
-	v := d.getGlobalProvider().Get(key)
-	if v.HasValue() {
-		str = v.String()
-	} else if str == "" {
-		return nil
+		return tmp
 	}
 
-	if value.IsNil() {
-		value.Set(reflect.New(value.Type()).Elem())
-	}
+	return v
+}
+
+// tryUnmarshallers checks if the value's type implements either one of standard interfaces in order:
+// 1. `json.Unmarshaler`
+// 2. `encoding.TextUnmarshaler`
+// 3. `yaml.Unmarshaler`
+// and tries it to populate the value.
+func (d *decoder) tryUnmarshalers(key string, value reflect.Value, def string) (bool, error) {
+	switch value.Kind() {
+	// value.IsNil panics if value.Kind() is not equal to one of these constants.
+	case reflect.Chan, reflect.Interface, reflect.Func, reflect.Map, reflect.Ptr:
+		if value.IsNil() {
+			value.Set(reflect.New(value.Type()).Elem())
+		}
 
-	// Value has to have a pointer receiver to be able to modify itself with TextUnmarshaller
-	if !value.CanAddr() {
-		return errorWithKey(errors.New("can't use TextUnmarshaller because value is not addressable"), key)
+	//TODO(alsam) Command-line provider implementation is made via maps, so we'll fail to merge slices and maps
+	// TestCommandLineProvider_NestedValues will fail in that case.
+	case reflect.Slice:
+		return false, nil
 	}
 
+	v := d.getGlobalProvider().Get(key)
 	switch t := value.Addr().Interface().(type) {
+	case json.Unmarshaler:
+		// Skip unmarshaling if there is no value.
+		if !v.HasValue() {
+			return true, nil
+		}
+
+		// Marshal the value first.
+		b, err := json.Marshal(jsonMap(v.Value()))
```

Comment:
it is an `interface{}`. We don't really know what a provider returns us: it can be a string, or it can be a `map[interface{}]interface{}`(in case of a yaml provider for structs).
The reason is we first doing a marshal first and unmarshal after is that we want to serialize this custom type first and then decode for example to a struct, basically doing a poor man's `mapstructure` implementation.